olmo/optim.py

```diff
@@ -141,10 +140,17 @@ def clip_grads_and_collect_metrics(
         def is_grad_norm_metric(metric_name: str) -> bool:
             return metric_name.startswith("grad/") and metric_name.endswith(".norm")
 
+        #######################################################################
+        # part 2: reduce metrics
+        #######################################################################
+        param_group_sharded = False
+        for group in self.param_groups:
+            param_group_sharded = param_group_sharded or group.get("sharded", False)
+
         # Now reduce metrics over all ranks.
         total_grad_norm: torch.Tensor
         per_param_avg_metrics: List[torch.Tensor] = []
-        if is_distributed():  # TODO (epwalsh): skip for non-sharded params
+        if is_distributed() and param_group_sharded:
```

Comment:
Will still need to reduce gradient metrics with non-sharded params since each rank will have different gradients.

Reply:
won't the gradients sync after loss.backward() call?

Reply:
Ah, yes you're right. My mistake

Reply:
in DDP loss.backward() syncs grads, optimizer.step() updates each copy with synced grads! Or am I getting this wrong? So, if loss.backward() syncs grads, then every rank must have the same grads

Reply:
@epwalsh does this apply to FSDP no_shard as well?